Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+# Automated Sharding
+
+* Owners:
+  * [Arthur Sens](https://github.com/ArthurSens)
+  * [Nicolas Takashi](https://github.com/nicolastakashi)
+* Related Tickets:
+  * [#4927](https://github.com/prometheus-operator/prometheus-operator/issues/4727)
+  * [#4946](https://github.com/prometheus-operator/prometheus-operator/issues/4946)
+  * [#4967](https://github.com/prometheus-operator/prometheus-operator/issues/4967)
+* Other docs:
+  * n/a
+
+This document aims to solve many small problems to unblock usage of Horizontal Pod Autoscalers with Prometheus sharding mechanism.
+
+# Why
+
+When operating hundreds/thousands of Prometheus instances that have very non-standardized loads, it becomes troublesome to fine-tune all of them. While several horizontal pod autoscalers have shown up in the market, they don't really work with Prometheus because they cause metrics duplication issues. Prometheus-Operator offers a sharding mechanism to solve this problem. Still, they also fall short when samples ingested per second change alongside the load on the monitored systems.
```

Comment:
I guess worth to mention this don't work because HPA use to scale the number of replicas from a Deployment/Statefulset and this will only increase reliability from Prometheus POV and not balance the load since all replicas will scrape all the same targets.

To solve that issue Prometheus should use a shard strategy, and then we can point to the shard document which explains Prometheus Operator sharding strategy.

wdyt?

Reply:
It is mentioned in "Pitfalls of current solution".  Did you miss it or you prefer that this is mentioned in the "Why" section?

Reply:
My feeling when I read the why is we're already introducing the Pitfalls.
Wdyt about this.

> Managing large Prometheus instances with highly variable workloads can be quite challenging. These instances often demand significant computing resources like memory and CPU. In many cases, the required resources exceed what operators have at their disposal. Additionally, these large Prometheus instances can take a long time to become operational after a restart due to the substantial Write-Ahead Log (WAL), which takes time to be replayed in memory.
>
> To address these issues, this document outlines a solution to enhance the Prometheus Operator's Sharding feature by automating horizontal scaling. This improvement aims to simplify the process of scaling Prometheus instances, making it more accessible and efficient for operators.

So we can left the Pitfalls explaining what it's doing currently.